### PR TITLE
Add documentation to functional_generic functions.

### PR DESCRIPTION
**Pull request by romCohen**

### CODE REVIEW

File: gamla/functional_generic.py [hunk 1]
```diff
@@ -11,6 +11,23 @@
 
 
 def compose_left(*funcs):
+    """Compose sync and async functions to operate in series.
+
+    Returns a function that applies other functions in sequence. The returned
+    function will be an async function iff atleast one of the functions in the
```
Comment (itayzit) [inline gamla/functional_generic.py:17]: at least

File: gamla/functional_generic.py [hunk 2]
```diff
@@ -66,6 +103,19 @@ async def async_composed(*args, **kwargs):
 
 
 def compose_sync(*funcs):
```
Comment (itayzit) [inline gamla/functional_generic.py:105]: I have a question not related to documentation: why isn't this a private function? we are not using it directly, and I don't think we would want to (so we can be agnostic about async).
wdyt?

File: gamla/functional_generic.py [hunk 3]
```diff
@@ -77,6 +127,23 @@ def composed(*args, **kwargs):
 
 
 def compose(*funcs):
+    """Compose sync and async functions to operate in series.
+
+    Returns a function that applies other functions in sequence. The returned
+    function will be an async function iff atleast one of the functions in the
```
Comment (itayzit) [inline gamla/functional_generic.py:133]: at least

File: gamla/functional_generic.py [hunk 4]
```diff
@@ -101,6 +168,25 @@ def reraise_and_log(e):
 
 
 def compose_many_to_one(incoming: Iterable[Callable], f: Callable):
+    """Compose an itterable of incoming sync and async functions to a single
+    sink function, in the same order as the incoming functions.
```
Comment (itayzit) [inline gamla/functional_generic.py:172]: I think this is not needed (doesn't add anything on top of what's written in 174-176)

File: gamla/functional_generic.py [hunk 5]
```diff
@@ -101,6 +168,25 @@ def reraise_and_log(e):
...
+
+    Returns a function that applies an itterable of other functions into a
+    single sink function. The returned function will be an async function iff
+    atleast one of the given functions is async.
+
+    The incoming functions are given the same input, and their output is mapped
+    to the coresponding input of the sink function, maintaining the incoming
+    function's order.
```
Comment (itayzit) [inline gamla/functional_generic.py:180]: again, I think lines 174-176 in enough, this is not needed

File: gamla/functional_generic.py [hunk 6]
```diff
@@ -173,6 +259,20 @@ def juxt(*args, **kwargs):
 
 
 def ternary(condition, f_true, f_false):
+    """Computes `f_true` with the givent arguments iff `condition` with the
+    same arguments return true, else Computes `f_false` with the givent
+    arguments.
+
+    Returns a function that computes `f_true` or `f_false` according to
+    `condition`. The returned function will be an async function iff
+    atleast one of the given functions is async.
```
Comment (itayzit) [inline gamla/functional_generic.py:268]: at least

File: gamla/functional_generic.py [hunk 7]
```diff
@@ -173,6 +259,20 @@ def juxt(*args, **kwargs):
 
 
 def ternary(condition, f_true, f_false):
+    """Computes `f_true` with the givent arguments iff `condition` with the
+    same arguments return true, else Computes `f_false` with the givent
```
Comment (itayzit) [inline gamla/functional_generic.py:263]: given*

File: gamla/functional_generic.py [hunk 8]
```diff
@@ -194,6 +294,18 @@ def ternary_inner(*args, **kwargs):
...
+#:  arguments.
+#:  Returns a function that computes `f_true` or `f_false` according to
+#:  `condition`. The returned function will be an async function iff
+#:  atleast one of the given functions is async.
+#:
+#:  >>> f = ternary(gamla.greater_than(5), gamla.identity, lambda i: -i)
+#:  >>> f(6)
+#:  '6'
+#:  >>> f(3)
+#:  '-3'
```
Comment (itayzit) [inline gamla/functional_generic.py:308]: ditto (at least, given)

File: gamla/functional_generic.py [hunk 9]
```diff
@@ -206,6 +318,20 @@ def unless(condition, f_false):
 
 
 def first(*funcs, exception_type: Type[Exception]):
+    """Constructs a function that computes all functions from `funcs`, the
+    function is async if at least one of the ginven functions is async. The
```
Comment (itayzit) [inline gamla/functional_generic.py:322]: given

File: gamla/functional_generic.py [hunk 10]
```diff
@@ -194,6 +294,18 @@ def ternary_inner(*args, **kwargs):
...
+#:  Returns a function that computes `f_true` or `f_false` according to
+#:  `condition`. The returned function will be an async function iff
+#:  atleast one of the given functions is async.
+#:
+#:  >>> f = ternary(gamla.greater_than(5), gamla.identity, lambda i: -i)
+#:  >>> f(6)
+#:  '6'
+#:  >>> f(3)
+#:  '-3'
 curried_ternary = ternary
```
Comment (itayzit) [inline gamla/functional_generic.py:309]: just talked to uri. I think we should delete it. wdyt?

**Review by itayzit — APPROVED** [review]

File: gamla/functional_generic.py (hunk 2 above)
Comment (romCohen) [reply gamla/functional_generic.py:105]: I think this is a good point, I'll delete

**Pull request merged**